### PR TITLE
Port most of the KV 1.3.2-1.4.0 updates

### PR DESCRIPTION
**Pull request by rzezeski**
This addresses most of #168. I still need to change the index list to be stored in cluster metadata so that Yokozuna can stop relying on ring events. That will be done as a separate PR rather than make this one bigger.

I performed a micro benchmark and there was no noticeable difference between before and after. Which is fine, I didn't necessarily make all the perf optimizations that were made in riak core as they are a bit tricker in Yokozuna.
- Use Riak KV's `get_index_n` instead of custom one.
- Add init/final to AAE compare API.
- Use `get_bucket/1` instead of `get_bucket\2` to make use of bucket cache.
- Add comment about AAE + ring resize (also see #279).

### CODE REVIEW

**Comment by hazen** [comment]
FWIW:

```
Test Results:
yz_wm_extract_test-bitcask   : pass
yz_stat_test-bitcask         : pass
yz_solr_start_timeout-bitcask: pass
yz_siblings-bitcask          : pass
yz_security-bitcask          : pass
yz_schema_admin-bitcask      : pass
yz_rs_migration_test-bitcask : pass
yz_pb-bitcask                : pass
yz_monitor_solr-bitcask      : pass
yz_mapreduce-bitcask         : pass
yz_languages-bitcask         : pass
yz_index_admin-bitcask       : pass
yz_fallback-bitcask          : pass
yz_errors-bitcask            : pass
yz_dt_test-bitcask           : pass
yokozuna_essential-bitcask   : pass
aae_test-bitcask             : pass
---------------------------------------------
0 Tests Failed
17 Tests Passed
That's 100.0% for those keeping score
```

File: src/yz_entropy_mgr.erl [hunk 1]
```diff
@@ -503,6 +503,13 @@ start_exchange(Index, Preflist, Ring, S) ->
     IsOwner = riak_core_ring:index_owner(Ring, Index) == node(),
     PendingChange = is_pending_change(Ring, Index),
 
+    %% NOTE: The riak_kv version of this function wraps this is a
+    %% try/catch to guard against the case where a ring size change
+    %% makes the index go way. I disagree with the use of try/catch as
```
Comment (coderoshi) [inline src/yz_entropy_mgr.erl]: "go away"

File: src/yz_entropy_mgr.erl [hunk 2]
```diff
@@ -503,6 +503,13 @@ start_exchange(Index, Preflist, Ring, S) ->
...
+    %% it could fail for other reaons. I believe this can be solved by
```
Comment (coderoshi) [inline src/yz_entropy_mgr.erl]: "reasons"

File: src/yz_exchange_fsm.erl [hunk 3]
```diff
@@ -216,8 +218,8 @@ repair(Partition, {_Reason, KeyBin}) ->
     %% match. In either case the object must be re-indexed.
     Ring = yz_misc:get_ring(transformed),
```
Comment (coderoshi) [inline src/yz_exchange_fsm.erl]: Ring isn't used until line 229, if it should be indexed. Perhaps delay initialization until after should_index true? Not really for micro optimization, just a coding practice thought.

File: src/yz_entropy_mgr.erl [hunk 4]
```diff
@@ -503,6 +503,13 @@ start_exchange(Index, Preflist, Ring, S) ->
     IsOwner = riak_core_ring:index_owner(Ring, Index) == node(),
     PendingChange = is_pending_change(Ring, Index),
 
+    %% NOTE: The riak_kv version of this function wraps this is a
```
Comment (jrwest) [inline src/yz_entropy_mgr.erl]: the problem is that when the ring resizes `riak_core_ring:index_owner` can `badmatch`. This code will still fail as written if the ring is resized (iirc).

Comment (rzezeski) [inline src/yz_entropy_mgr.erl]: Good call. I misread this code and let my snark of try/catch blind me. Porting this change now.

**Comment by rzezeski** [comment]
Just so I don't forget I found an issue while running dialyzer:

```
yz_exchange_fsm.erl:203: Call to missing or unexported function yz_kv:get_index/2
```

It's a simple fix but think I might take the time tomorrow to add to aee_test so that this code path is tested.

**Comment by coderoshi** [comment]
+1

```
yz_wm_extract_test-bitcask   : pass
yz_stat_test-bitcask         : pass
yz_solr_start_timeout-bitcask: pass
yz_siblings-bitcask          : pass
yz_security-bitcask          : pass
yz_schema_admin-bitcask      : pass
yz_rs_migration_test-bitcask : pass
yz_pb-bitcask                : pass
yz_monitor_solr-bitcask      : pass
yz_mapreduce-bitcask         : pass
yz_languages-bitcask         : pass
yz_index_admin-bitcask       : pass
yz_fallback-bitcask          : pass
yz_errors-bitcask            : pass
yz_dt_test-bitcask           : pass
yokozuna_essential-bitcask   : pass
aae_test-bitcask             : pass
---------------------------------------------
0 Tests Failed
17 Tests Passed
That's 100.0% for those keeping score
```

**Comment by rzezeski** [comment]
@coderoshi Sorry but I was still working on this PR. I suppose my last
last comment wasn't very clear. Do you mind running the entire suite
against latest version of this branch one more time?

FWIW, here's a 100% test run I got after the above changes.

```
Test Results:
yz_wm_extract_test-bitcask   : pass
yz_stat_test-bitcask         : pass
yz_solr_start_timeout-bitcask: pass
yz_siblings-bitcask          : pass
yz_security-bitcask          : pass
yz_schema_admin-bitcask      : pass
yz_rs_migration_test-bitcask : pass
yz_pb-bitcask                : pass
yz_monitor_solr-bitcask      : pass
yz_mapreduce-bitcask         : pass
yz_languages-bitcask         : pass
yz_index_admin-bitcask       : pass
yz_fallback-bitcask          : pass
yz_errors-bitcask            : pass
yz_dt_test-bitcask           : pass
yokozuna_essential-bitcask   : pass
aae_test-bitcask             : pass
---------------------------------------------
0 Tests Failed
17 Tests Passed
That's 100.0% for those keeping score
```

I get the following dialyzer output which looks the same as before the
patch. I plan to address the dialyzer issues in a separate PR.

```
yokozuna.erl:37: The call application:set_env('yokozuna',{'component',_},'false') breaks the contract (Application,Par,Val) -> 'ok' when is_subtype(Application,atom()), is_subtype(Par,atom()), is_subtype(Val,term())
yokozuna.erl:44: The call application:set_env('yokozuna',{'component',_},'true') breaks the contract (Application,Par,Val) -> 'ok' when is_subtype(Application,atom()), is_subtype(Par,atom()), is_subtype(Val,term())
yz_index.erl:233: The call riak_core_util:rpc_every_member_ann('yz_index','reload_local',[any(),...],TO::any()) will never return since it differs in the 2nd argument from the success typing arguments: (atom() | tuple(),fun(),[any()],'infinity' | non_neg_integer())
yz_kv.erl:92: The call lists:sort(Vclock::vclock:vclock()) contains an opaque term as 1st argument when a structured term of type [any()] is expected
yz_kv.erl:118: Invalid type specification for function yz_kv:get_tree_build_time/1. The success typing is (hashtree:hashtree()) -> any()
yz_kv.erl:182: The call yz_kv:dont_index(Obj::riak_object:riak_object(),Reason::any(),P::non_neg_integer(),ShortPL::{non_neg_integer(),pos_integer()},BKey::{binary(),binary()}) breaks the contract (obj(),write_reason(),p(),bkey(),short_preflist()) -> 'ok'
yz_pb_admin.erl:32: Callback info about the riak_api_pb_service behaviour is not available
yz_pb_search.erl:30: Callback info about the riak_api_pb_service behaviour is not available
Unknown types:
  orddict:orddict/2
  xmerl_scan:document/0
```

**Comment by rzezeski** [comment]
And one last update. Had to bring in latest development.

Once again tests passed.

```
Test Results:
yz_wm_extract_test-bitcask   : pass
yz_stat_test-bitcask         : pass
yz_solr_start_timeout-bitcask: pass
yz_siblings-bitcask          : pass
yz_security-bitcask          : pass
yz_schema_admin-bitcask      : pass
yz_rs_migration_test-bitcask : pass
yz_pb-bitcask                : pass
yz_monitor_solr-bitcask      : pass
yz_mapreduce-bitcask         : pass
yz_languages-bitcask         : pass
yz_index_admin-bitcask       : pass
yz_fallback-bitcask          : pass
yz_errors-bitcask            : pass
yz_dt_test-bitcask           : pass
yokozuna_essential-bitcask   : pass
aae_test-bitcask             : pass
---------------------------------------------
0 Tests Failed
17 Tests Passed
That's 100.0% for those keeping score
```

**Comment by coderoshi** [comment]
Take two, still +1

```
yz_wm_extract_test-bitcask   : pass
yz_stat_test-bitcask         : pass
yz_solr_start_timeout-bitcask: pass
yz_siblings-bitcask          : pass
yz_security-bitcask          : pass
yz_schema_admin-bitcask      : pass
yz_rs_migration_test-bitcask : pass
yz_pb-bitcask                : pass
yz_monitor_solr-bitcask      : pass
yz_mapreduce-bitcask         : pass
yz_languages-bitcask         : pass
yz_index_admin-bitcask       : pass
yz_fallback-bitcask          : pass
yz_errors-bitcask            : pass
yz_dt_test-bitcask           : pass
yokozuna_essential-bitcask   : pass
aae_test-bitcask             : pass
---------------------------------------------
0 Tests Failed
17 Tests Passed
That's 100.0% for those keeping score
```

**Pull request merged**